python/ray/_private/utils.py

```diff
@@ -801,6 +802,97 @@ def set_sigterm_handler(sigterm_handler):
         signal.signal(signal.SIGTERM, sigterm_handler)
 
 
+_signal_handler_installed = False
+_graceful_shutdown_in_progress = False
+
+
+def install_driver_signal_handler() -> None:
+    """Install SIGTERM handler for Ray driver processes.
+
+    Implements graceful-then-forced shutdown semantics:
+    - First SIGTERM: trigger graceful shutdown via sys.exit() (allows atexit handlers)
+    - Second SIGTERM: escalate to immediate forced shutdown via os._exit(1)
```

Comment:
`SIGTERM` handler should be idempotent. There's no purpose to having an "escalation" here because the caller can & should `SIGKILL` if they want to un-gracefully kill the process.